[ONNX] Relax unsupported node analysis on complex dtype

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113785

In cases like #113444, users usually stop at UnsupportedNodeAnalysis with unsupported nodes information. Although in SARIF, they can clearly see it's due to lack of COMPLEX support, in screen error message, it's only showing original FX node name, such as `aten.mul.Tensor`. ~~This PR catches the information from diagnostic messages and reveal it to users.~~

The root cause is that UnsupportedNodeAnalysis is leveraging on `onnxfunction_dispatcher.get_function_overloads()` to decide if an ATen is supported or not. However, in `onnxfunction_dispatcher.get_function_overloads()`, lacking of complex function support is considered unsupported. This PR defines Unsupported FX nodes as not in registry.

titaiwangms committed:
[ONNX] Specify unsupported data type in unsupported node analysis

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/analysis/unsupported_nodes.py:
```diff
@@ -61,14 +63,23 @@ def analyze(
-            target = node.target
-            op_to_target_mapping.setdefault(op, {}).setdefault(str(target), None)
+                    # Extract the dtype information from the diagnostic message.
+                    if e.diagnostic.message and e.diagnostic.message.startswith(
```
can we raise specific exception type instead of string matching to detect this error? string matching on logs is usually flaky. we log in one place and check the string matching in another. Unless we have a test for guarding this string matching, it will eventually break

titaiwangms replied:
We all know that. The root cause is that the mismatch message in SARIF and screen error message. I wonder if we should create a new rule to get a new template in this error message. @BowenBao Do you have any suggestion.

titaiwangms replied:
Added test for the current approach

BowenBao replied:
Maybe we can just expand the error being emitted from this analysis, showing `diagnostic.message` for each op. That way we don't need to do string parsing on them.

BowenBao replied:
Or maybe we can make it standard to attach dtype with optype, through looking at the graph instead of string parsing.

titaiwangms replied:
I see. The first one seems good and we can align the message in screen printing with SARIF.

thiagocrepaldi replied:
> Maybe we can just expand the error being emitted from this analysis, showing `diagnostic.message` for each op. That way we don't need to do string parsing on them.

Not relying on string matching is a nice improvement. Could we go further and populate `op_to_target_mapping` (thus the error message) without using diagnostic at all?

The diagnostics is a passive component that captures extra debug information and put on a SARIF file - only that. Ideally, it should not influence (by populating `op_to_target_mapping`) and alter `UnsupportedFxNodesAnalysisResult` outcome. 

Maybe I am missing something, but by reusing diagnostics for anything other than SARIF, we are creating a direct coupling between `UnsupportedFxNodesAnalysis` and SARIF; That is, `UnsupportedFxNodesAnalysis` cannot properly report the error without depending on diagnostics machinery.

IMO, if we could find a way to capture the missing information on the analysis pass independently of diagnostics, we can keep the analysis pass and diagnostics decoupled.

titaiwangms replied:
Made some modifications. Eseentially, now we don't leverage on `get_function_overloads`, instead, we use `is_registered_op` to capture the ATen that doesn't even have at least one ONNX invariant. Completely extract out the arg/function sig matching from UnsupoortedNodeAnalysis.

If ATen is missing complex support, the error will be emitted during dispatcher now, which provides a more suitable error message.

titaiwangms committed:
Update on "[ONNX] Specify unsupported data type in unsupported node analysis"


In cases like #113444, users usually stop at UnsupportedNodeAnalysis with unsupported nodes information. Although in SARIF, they can clearly see it's due to lack of COMPLEX support, in screen error message, it's only showing original FX node name, such as `aten.mul.Tensor`. This PR catches the information from diagnostic messages and reveal it to users.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Specify unsupported data type in unsupported node analysis"


In cases like #113444, users usually stop at UnsupportedNodeAnalysis with unsupported nodes information. Although in SARIF, they can clearly see it's due to lack of COMPLEX support, in screen error message, it's only showing original FX node name, such as `aten.mul.Tensor`. This PR catches the information from diagnostic messages and reveal it to users.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
